### ISSUE TITLE
featurize_sequence and logit values

### ISSUE THREAD

**Issue by rnyak** [IndicoDataSolutions/finetune#447]
**Is your feature request related to a problem? Please describe.**

Hi @madisonmay,

Following up my previous issue and your response below:
```
I'll roll this into our documentation shortly, but the gist is that the sequence_features() method provides you with access to the output of the final transformer layer before pooling.

Example:

from finetune import Classifier
model = Classifier()
sequence_features = model.featurize_sequence(["Featurize sequence outputs a vector per token"])
print(sequence_features.shape)
# outputs (1, 512, 768) or more generically (n_examples, max_length, hidden_dim)
```
Not sure I see a pooling layer. Did you mean before perceptron or dropout? See the function below:
```
def classifier (...):

   ...

 with tf.variable_scope("classifier", reuse=reuse):
        hidden = dropout(hidden, config.clf_p_drop, train)
        clf_logits = perceptron(hidden, n_targets, config)
        if targets is None:
            clf_losses = None
        else:
            clf_losses = tf.nn.softmax_cross_entropy_with_logits_v2(
                logits=clf_logits, labels=tf.stop_gradient(targets)
            )

            clf_losses = _apply_class_weight(
                clf_losses, targets, kwargs.get("class_weights")
            )

            # From Unsupervised Data Augmentation for Consistency Training, Xie et al. 2019
            if config.tsa_schedule:
                clf_logits, clf_losses = tsa_loss(
                    n_targets, config, clf_losses, clf_logits, targets
                )

        return {"logits": clf_logits, "losses": clf_losses}
```

**Describe the solution you'd like**

What I would like to get is the "hidden" layer values before calling perceptron model, and  clf_logits  as output of linear perceptron model. 

The hidden param in the classifier function is defined as `:param hidden: The output of the featurizer. [batch_size, embed_dim].`  I use featurizer after finetuning the model, with an inference set, and I get `[n_samples, embed_dim]` as output from `model.featurize (inferenceX)` where inferenceX is an array of text samples. If my inferenceX set has 10000 samples, then, I get, [10000, 768], whereas  model.featurize_sequence(inferenceX) output is [n_samples, max_length, embed_dim], i.e. [10000, 1024, 768].

Does the model.featurize(samples) correspond to the "hidden" in the classifier function above?

**Comment by benleetownsend** [comment]
Sorry for the slow response, `model.featurize(samples)` does correspond to the hidden above.

If I understand correctly you also want to get the logits we do not have it exposed to directly get logits, but you can get probabilities with `model.predict_proba`.

**Comment by rnyak** [comment]
Yes, I am getting the probas, but wanted to have logits also. and the last layer values before generating logits.

**Comment by benleetownsend** [comment]
Would it be sufficient to have log probabilities in replacement for logits? If not you'd have to write some custom code for this to expose the logits. I'm afraid returning multiple things in the same call is not currently supported so you'd either have to do this in two model calls or this would require some code changes.

To get logits is probably as simple as

```
from finetune import Classifier

class LogitClassifier(Classifier):
    def _predict_proba_op(self, logits, **kwargs):
        return logits
```
Now when you call `LogitClassifier.predict` you'd get logits instead of probabilities.

To add features alongside the logits you'd want to do something like this. Caveat is that chunk_long_sequences has to be turned off for this to function, if you require longer sequence lengths then this may be a problem.
```
from finetune import Classifier
from finetune.model import PredictMode

class LogitClassifierWithFeats(Classifier):
    def _predict_proba_op(self, logits, **kwargs):
        return logits

    def predict_with_feats(self, Xs, context=None):
        raw_preds = self._inference(Xs, predict_keys=[PredictMode.PROBAS, PredictMode.FEATS], context=context)
        classes = self.input_pipeline.label_encoder.classes_
        formatted_predictions = []
        features = []
        for pred in raw_preds:
            probas = pred[PredictMode.PROBAS]
            feats = pred[PredictMode.FEATS]
            formatted_predictions.append(dict(zip(classes, probas)))
            features.append(feats)
        return formatted_predictions
```

**Comment by benleetownsend** [comment]
Did this work out?

**Comment by rnyak** [comment]
@benleetownsend did not try it yet. Thanks for following up. I am closing this issue.
